engine/src/main/java/org/hibernate/search/util/impl/FileHelper.java

```diff
@@ -84,58 +84,50 @@ else if ( !destination.isDirectory() ) {
 		}
 	}
 
-	public static void synchronize(File source, File destination, boolean smart) throws IOException {
+	public static void synchronize(Path source, Path destination, boolean smart) throws IOException {
 		synchronize( source, destination, smart, DEFAULT_COPY_BUFFER_SIZE );
 	}
 
-	public static void synchronize(File source, File destination, boolean smart, long chunkSize) throws IOException {
+	public static void synchronize(Path source, Path destination, boolean smart, long chunkSize) throws IOException {
 		if ( chunkSize <= 0 ) {
 			log.checkSizeMustBePositive();
 			chunkSize = DEFAULT_COPY_BUFFER_SIZE;
 		}
-		if ( source.isDirectory() ) {
-			if ( !destination.exists() ) {
-				if ( !destination.mkdirs() ) {
-					throw new IOException( "Could not create path " + destination );
-				}
-			}
-			else if ( !destination.isDirectory() ) {
-				throw new IOException(
-						"Source and Destination not of the same type:"
-								+ source.getCanonicalPath() + " , " + destination.getCanonicalPath()
-				);
-			}
-			String[] sources = source.list();
-			Set<String> srcNames = new HashSet<String>( Arrays.asList( sources ) );
-			String[] dests = destination.list();
+		if ( Files.isDirectory( source ) ) {
+			Files.createDirectories( destination );
+			Set<Path> sources = Files.list( source ).collect( Collectors.toSet() );
```

Comment:
Good catch. I assumed this was handled by the Stream/Collector framework but apparently it will only do that if you flatmap the stream (?!)